### DISCUSSION TITLE
Discord Auth requests doubled in Chrome

### DISCUSSION THREAD

**Discussion by whois-jon-peterson** [requarks/wiki#6349] [Error / Bug Report] [answered]
Wiki.js version 2.5.286 running [via Docker](https://docs.requarks.io/install/ubuntu)
visiting using Chrome 112.0.5615.49 (stable) on Windows 11

I am experiencing an issue with Discord logins failing due to rate limits fetching guilds (servers). After a bit of digging, I found that in Chrome (but not in Firefox) the auth flow was making two requests to the callback endpoint within 200-300ms of one another, the first of which is always canceled.
![image](https://user-images.githubusercontent.com/3322920/231261969-0ac36a9c-dd76-40be-872e-a0d12851e15f.png)
_Note: I've masked the `code` params in this screenshot, but they're each distinct._

This results in the server receiving and handling two separate parallel requests to `/login/:strategy/callback` and thus the DiscordStrategy auth provider. This makes calls to two Discord endpoints (`/users/@me` and `/users/@me/guilds`), and rate limits on the the latter are too strict for this (1 request per 2 seconds), so the second auth attempt fails. Unfortunately, this is the one that _isn't_ canceled, so it's what the user gets a response from.

I have so far been unable to determine why there are two requests to the callback endpoint, much less why it's only happening in Chrome. I'm currently working around this by modifying the `passport-discord` lib to add an in-memory cache to track open auth attempts (keyed on `accessToken`) and store retrieved profile data for a few seconds so I can avoid calling out to Discord on the second auth attempt, and instead just wait for the first to complete.

An alternative workaround would be to remove the constraint on Server ID from the admin panel's authentication settings, and remove the `guilds` scope from {embedded below} as this would skip the rate-limited request altogether... however I do really want that constraint in place for the wiki instance I'm running. https://github.com/requarks/wiki/blob/42d5fd6cb898f1574b631c6c3dcd6cbae40b541c/server/modules/authentication/discord/authentication.js#L13-L19

The optimal solution of course, would be to figure out why there's two requests to the callback endpoint and prevent that from happening. I'm afraid I'm out of my depth on the front end though.

**Comment by Facinorous-420** [comment]
I am also facing this issue on WikiJS 2.5.298 running in a docker container. Hopefully we can get to the bottom of this as none of my users can log in currently.

**Comment by whois-jon-peterson** [comment, marked as answer]
I'm beginning to suspect the Discord auth API itself was at fault here. The issue has gone away. My cache patch was preventing the issue from causing problems, but you could still see the doubled callback in the network tab of Chrome dev tools... and that's gone now with no further changes on my part.

**Comment by Facinorous-420** [comment]
Yeah seems to have also resolved itself on my side too. Thanks for reporting back
